### ISSUE TITLE
Proposal: Cleaner JSON Schemas

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#3625]
I've been thinking about this for a bit, and I kind of hate what we have to do to support optional configuration options, like this:

https://github.com/eslint/eslint/blob/232075345f23131bcbb8013c1ed3f7c5708f3095/lib/rules/no-restricted-syntax.js#L36

IMHO, needing to put the error codes into the rules, explicitly, is a barrier to having a clean and loosely-coupled API.

Therefore, I'd like to propose that we no longer require rules to include `enum: [1, 2, 3]` in any case and that the rule schema should, instead, only refer to the values provided in `context.options`.

How would this work?

Basically, the validator will always just `shift()` off the first option (the severity) and manually validate that number. We already have a special case to provide a better message (https://github.com/eslint/eslint/blob/master/lib/config-validator.js#L83), but really, we don't need JSON schema to determine if a value is 0, 1, or 2. We can just do that ourselves.

So we remove the severity from the array and *then* apply the JSON schema.

The end result is that no schema ever needs to worry about severity codes under any circumstances.

Thoughts?

**Comment by ilyavolodin** [comment]
It looks like linked example from `no-restricted-syntax` is done differently then most of our schemas. In most places we don't do schema as an object, we do it as an array, and this [line of code](https://github.com/eslint/eslint/blob/master/lib/config-validator.js#L38) takes care of the error levels for us.
So I think currently no rule schema should ever need to worry about severity code, unless the person who created that schema really wants to do that, for some reason.

**Comment by nzakas** [comment]
That was just one example. Here's another:

https://github.com/eslint/eslint/blob/9bc543fe221f91fefc894ccbca2b0860a70475db/lib/rules/quote-props.js#L187

This was recommended for whenever rules can have varying numbers of options. I think this needs to be addressed so it doesn't continue to be used.

**Comment by ilyavolodin** [comment]
Oh, I see now. That's done so that you could do
```json
    "uniqueItems": true,
    "minItems": 1
```
Sorry, I missed that part. In that case I agree.

**Comment by BYK** [comment]
:+1:

**Comment by btmills** [comment]
:+1: This design is much cleaner. I see no reason why it wouldn't work.

**Comment by nzakas** [comment]
Cool, I'll add this to the 2.0.0 roadmap.

**Comment by gyandeeps** [comment]
If I understand this proposal correctly then is it really a breaking change?

**Comment by ilyavolodin** [comment]
It would be a breaking change only for plugin creators, consumers will not see anything at all.

**Comment by gyandeeps** [comment]
Got it. Thanks @ilyavolodin

**Comment by nzakas** [comment]
I'll take a stab at this, I think it should be fairly easy.
